packages/ui/src/components/va-tabs/hooks/useTabsView.ts

```diff
@@ -0,0 +1,414 @@
+        ) {
+          offsetToSet = tabsService.value.tabs[i].leftSidePosition
+          break
+        }
+      }
+    }
+
+    tabsContentOffset.value = Math.max(0, offsetToSet)
+
+    emit('click:prev')
+  }
+
+  const movePaginationRight = () => {
+    const containerClientWidth = getClientWidth(container.value)
+    const containerRightSide = tabsContentOffset.value + containerClientWidth
+    let offsetToSet = containerRightSide
+
+    if (tabsService.value) {
+      for (let i = 0; i < tabsService.value.tabs.length; i++) {
+        if (tabsService.value.tabs[i].rightSidePosition > containerRightSide) {
+          offsetToSet = tabsService.value.tabs[i].leftSidePosition
+          if (tabsContentOffset.value < offsetToSet) {
+            break
+          }
+        }
+      }
+    }
+
+    const lastTab = tabsService.value?.tabs[tabsService.value.tabs.length - 1]
+    const maxOffset = (lastTab?.rightSidePosition || 0) - containerClientWidth
+
+    offsetToSet = Math.min(maxOffset, offsetToSet)
+    tabsContentOffset.value = Math.max(0, offsetToSet)
+
+    emit('click:next')
+  }
+
+  const updateSlider = (tab: TabComponent) => {
+    // @ts-ignore
+    const tabOffsetTop = tab.tabElement?.offsetTop || 0
+    // @ts-ignore
+    const tabOffsetLeft = tab.tabElement?.offsetLeft || 0
+    // @ts-ignore
+    const tabClientHeight = tab.tabElement?.clientHeight || 0
+    // @ts-ignore0
+    const tabClientWidth = tab.tabElement?.clientWidth || 0
+
+    if (props.vertical) {
+      const containerClientHeight = container.value?.clientHeight || 0
+      const calculatedSliderOffsetY = containerClientHeight - tabOffsetTop - tabClientHeight
+
+      sliderOffsetY.value = Math.max(calculatedSliderOffsetY, 0)
+      sliderHeight.value = tabClientHeight
+      sliderOffsetX.value = 0
+      sliderWidth.value = 0
+    } else {
+      sliderOffsetX.value = tabOffsetLeft
+      sliderWidth.value = tabClientWidth
+      sliderOffsetY.value = 0
+      sliderHeight.value = null
+    }
+  }
+
+  const includeAnimation = () => {
+    if (!animationIncluded.value) {
+      requestAnimationFrame(() => {
+        animationIncluded.value = true
+      })
+    }
+  }
+
+  const redrawTabs = () => {
+    const oldShowPaginationValue = showPagination.value
+
+    updatePagination()
+
+    if (oldShowPaginationValue === showPagination.value) {
+      updateTabsState()
+      includeAnimation()
+    } else {
+      requestAnimationFrame(() => {
+        updateTabsState()
+        includeAnimation()
+      })
+    }
+  }
+
+  const selectTab = (tab: TabComponent) => {
+    if (tab) {
+      valueComputed.value = tab.name || tab.id
+
+      if (props.stateful) {
+        updateTabsState()
+      }
+    }
+  }
+
+  provide(TabsViewKey, {
+    disabled: props.disabled,
+    selectTab,
+    moveToTab,
+  })
+
+  tabsService.value = useTabsService({
+    parentDisabled: props.disabled,
+    selectTab,
+    moveToTab,
+  })
```

Comment:
There's no reason for tabsService to be a ref. 

```suggestion
  const tabsService = useTabsService({
    parentDisabled: props.disabled,
    selectTab,
    moveToTab,
  })
```

Reply:
Components logic, I refactored without any separate methods.